### PR TITLE
Backend JWT token generation support within microgateway

### PR DESCRIPTION
**Pull request by HeshanSudarshana**
### Purpose
Provides the capability of generating backend JWT tokens within microgateway with the ability to customize the population of claims and the generation of tokens.

### Issues
Fixes https://github.com/wso2/product-microgateway/issues/1196

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes
### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [ ] Assigned 'Type' label
- [ ] Assigned the project
- [ ] Validated respective github issues
- [ ] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-core/pom.xml [hunk 1]
```diff
@@ -124,6 +124,21 @@
             <groupId>org.wso2.carbon.callhome</groupId>
             <artifactId>core</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.nimbusds</groupId>
+            <artifactId>nimbus-jose-jwt</artifactId>
+            <version>8.8</version>
```
Comment (Rajith90) [inline components/micro-gateway-core/pom.xml:130]: Move the versions to the portperties

File: distribution/resources/conf/default-micro-gw.conf.template [hunk 2]
```diff
@@ -296,3 +296,42 @@
...
+  privateKeyAlias="ballerina"
+  # JWT token expiry time - ms (valid only if the jwt generator caching mechanism is disabled)
+  tokenExpiry=900000
+  # Restricted claims as an array that should not be included in the backend JWT token
+  # Example: restrictedClaims=["claim1","claim2","claim3"]
+  restrictedClaims=[]
+  # Token issuer standard claim
+  issuer="wso2.org/products/am"
+  # Token audience standard claim
+  audience="http://org.wso2.apimgt/gateway"
```
Comment (Rajith90) [inline distribution/resources/conf/default-micro-gw.conf.template:320]: Audience should be supported for multiple values

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/jwt_auth_handler.bal [hunk 3]
```diff
@@ -62,22 +143,192 @@ public type JWTAuthHandler object {
...
-        if (customClaims is map<json> && customClaims.hasKey(BACKEND_JWT)) {
-            printDebug(KEY_JWT_AUTH_PROVIDER, "Set backend jwt header.");
-            req.setHeader(jwtheaderName, customClaims.get(BACKEND_JWT).toString());
+# Identify the api details from the subscribed apis in the authentication token.
+#
+# + payload - The payload of the authentication token
+# + apiName - name of the current API
+# + apiVersion - version of the current API
+# + return - Returns map<string> with the extracted details.
+public function checkSubscribedAPIs(jwt:JwtPayload payload, string apiName, string apiVersion) returns map<string> {
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/jwt_auth_handler.bal:168]: Change the metod name. This method resolves the selected API details

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    }
+
+    /**
+     * Used to get the expiration time of the token
+     */
+    public long getTTL() {
+        if (ttl != -1) {
+            return ttl;
+        }
+        synchronized (AbstractMGWJWTGenerator.class) {
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:267]: Re check why we need synchronize block here

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 5]
```diff
@@ -0,0 +1,366 @@
+package org.wso2.micro.gateway.core.jwtgenerator;
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:1]: License header is missing

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+import java.security.PrivateKey;
+import java.security.Signature;
+import java.security.cert.Certificate;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *  Abstract class for generate JWT for backend claims
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:21]: Checkstyle failure if no `.` at the end of first sentence.
```suggestion
 *  Abstract class for generate JWT for backend claims.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+    public int getJwtExpiryTime() {
+        return jwtExpiryTime;
+    }
+
+    public void setJwtExpiryTime(int jwtExpiryTime) {
+        this.jwtExpiryTime = jwtExpiryTime;
+    }
+
+    /**
+     * Used to generate the JWT token
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:185]:
```suggestion
     * Used to generate the JWT token.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+            }
+            String base64UrlEncodedAssertion = encode(signedAssertion);
+            return base64UrlEncodedHeader + '.' + base64UrlEncodedBody + '.' + base64UrlEncodedAssertion;
+        } else {
+            return base64UrlEncodedHeader + '.' + base64UrlEncodedBody + '.';
+        }
+    }
+
+    /**
+     * Used to build the JWT header
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:213]:
```suggestion
     * Used to build the JWT header.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+        if (NONE.equals(signatureAlgorithm)) {
+            StringBuilder jwtHeaderBuilder = new StringBuilder();
+            jwtHeaderBuilder.append("{\"typ\":\"JWT\",");
+            jwtHeaderBuilder.append("\"alg\":\"");
+            jwtHeaderBuilder.append("none");
+            jwtHeaderBuilder.append('\"');
+            jwtHeaderBuilder.append('}');
+
+            jwtHeader = jwtHeaderBuilder.toString();
+
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:226]:
```suggestion
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+        } else if (SHA256_WITH_RSA.equals(signatureAlgorithm)) {
+            jwtHeader = addCertToHeader();
+        }
+        return jwtHeader;
+    }
+
+    /**
+     * Used to sign the JWT using the keystore
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:234]:
```suggestion
     * Used to sign the JWT using the keystore.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+     */
+    public byte[] signJWT(String assertion) throws Exception {
+        FileInputStream is;
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:237]: better to add some debug logs in each step of JWT generation to find out where it fails if we notice an error in generation.

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+
+        // close the file stream
+        is.close();
+
+        //sign the assertion and return the signature
+        return signature.sign();
+    }
+
+    /**
+     * Used to get the expiration time of the token
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:261]:
```suggestion
     * Used to get the expiration time of the token.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+                ttl = cacheExpiry;
+            } else {
+                ttl = jwtExpiryTime;
+            }
+            return ttl;
+        }
+    }
+
+    /**
+     * Used to add "ballerina"the certificate from the keystore to the header
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:281]:
```suggestion
     * Used to add "ballerina"the certificate from the keystore to the header.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+        jwtHeader.append('}');
+
+        // close the file stream
+        is.close();
+
+        return jwtHeader.toString();
+    }
+
+    /**
+     * Used to build the body with claims
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:321]:
```suggestion
     * Used to build the body with claims.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+        }
+        for (Map.Entry<String, Object> claimEntry : claims.entrySet()) {
+            jwtClaimSetBuilder.claim(claimEntry.getKey(), claimEntry.getValue());
+        }
+        JWTClaimsSet jwtClaimsSet = jwtClaimSetBuilder.build();
+        return jwtClaimsSet.toJSONObject().toString();
+    }
+
+    /**
+     * Used for base64 encoding
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:344]:
```suggestion
     * Used for base64 encoding.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+     */
+    public String encode(byte[] stringToBeEncoded) {
+        return java.util.Base64.getUrlEncoder().encodeToString(stringToBeEncoded);
+    }
+
+    //Helper method to hexify a byte array
+    public String hexify(byte bytes[]) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/AbstractMGWJWTGenerator.java:351]: Do we need this as public method?

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorImpl.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
+package org.wso2.micro.gateway.core.jwtgenerator;
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorImpl.java:1]: License header missing

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorImpl.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ *  Class to implement standard claims and custom claims
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorImpl.java:13]:
```suggestion
 *  Class to implement standard claims and custom claims.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java [hunk 19]
```diff
@@ -0,0 +1,115 @@
+package org.wso2.micro.gateway.core.jwtgenerator;
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java:1]: Add license header

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+            return true;
+        } catch (InstantiationException | IllegalAccessException | ClassNotFoundException
+                    | InvocationTargetException | NoSuchMethodException e) {
+            log.error("Error while loading the jwt generator class: " + className, e);
+        }
+        return false;
+    }
+
+    /**
+     * Invoke token generation method
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java:56]:
```suggestion
     * Invoke token generation method.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+        abstractMGWJWTGenerator.setApiDetails(apiDetailsMap);
+        return abstractMGWJWTGenerator.generateToken(jwtInfoMap);
+    }
+
+    public static String invokeGetKeystorePath(String fullPath) {
+        return getKeyStorePath(fullPath);
+    }
+
+    /**
+     * Convert ArrayValue to Array
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java:70]:
```suggestion
     * Convert ArrayValue to Array.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+                array[i] = convertArrayValueToArray((ArrayValue) arrayValue.get(i));
+            } else {
+                array[i] = arrayValue.get(i);
+            }
+        }
+        return array;
+    }
+
+    /**
+     * Convert MapValue to Map
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java:87]:
```suggestion
     * Convert MapValue to Map.
```

File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+                map.put(key.toString(), convertArrayValueToArray(arrayValue));
+            } else {
+                map.put(key.toString(), valueObject);
+            }
+        }
+        return map;
+    }
+
+    /**
+     * Used to get the keystore path
```
Comment (praminda) [inline components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/jwtgenerator/MGWJWTGeneratorInvoker.java:108]:
```suggestion
     * Used to get the keystore path.
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/utils/utils.bal [hunk 24]
```diff
@@ -126,7 +126,7 @@ public function isAccessTokenExpired(APIKeyValidationDto apiKeyValidationDto) re
     if (issueTime is string) {
         issuedTime = 'int:fromString(issueTime);
     }
-    int timestampSkew = getConfigIntValue(KM_CONF_INSTANCE_ID, TIMESTAMP_SKEW, DEFAULT_TIMESTAMP_SKEW);
+    int timestampSkew = getConfigIntValue(SERVER_CONF_ID, TIMESTAMP_SKEW, DEFAULT_TIMESTAMP_SKEW);
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/utils/utils.bal:129]: Backward compatibility should be supported

File: tests/src/test/java/org/wso2/micro/gateway/tests/jwtGenerator/JWTGenerationTestCase.java [hunk 25]
```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (praminda) [inline tests/src/test/java/org/wso2/micro/gateway/tests/jwtGenerator/JWTGenerationTestCase.java:2]:
```suggestion
 * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: tests/src/test/resources/confs/base.conf [hunk 26]
```diff
@@ -138,3 +137,6 @@ name = "jaeger"
 
 [b7a.observability.tracing.jaeger.reporter]
 port=5775
+
+[server]
+  timestampSkew=5000
```
Comment (praminda) [inline tests/src/test/resources/confs/base.conf:142]:
```suggestion
  timestampSkew=5000

```

**Pull request closed without merging**